Fixed #31180 -- Configured applications automatically.

aaugustin opened the pull request:
Many pluggable applications are configured with a dotted path to the app (e.g. `"django.contrib.admin"`) in `INSTALLED_APPS`.

The application then relies on `default_app_config` for locating its `AppConfig` class (e.g. `"django.contrib.admin.apps.AdminConfig"`).

`default_app_config` was intended to allow author of pluggable applications to take advantage of `AppConfig` features without require their users to change `INSTALLED_APPS`, in order to increase adoption.

Unfortunately, that reduced the incentive for directly putting paths to `AppConfig` in `INSTALLED_APPS`. Even though that was the official recommendation, it didn't gain traction.

If I had better anticipated this result, I wouldn't have introduced `default_app_config` in the app-loading refactor in Django 1.7.

We've spent about five years failing at getting people to adopt explicit configuration. It's time to admit failure. Practicality beats purity.

With the benefit of hindsight, I'm proposing to load `AppConfig` classes from a conventional location, namely the `apps` submodule inside an application.

This is inconsistent with Django's design philosophy, as Django favors configuration over convention, but not much worse than the `default_app_config` convention.

There's a small risk of backwards-incompatibility. Django could import an `apps.py` submodule designed for another purpose and that could have side effects. I think that risk is low.

aaugustin commented:
django-developers discussion: https://groups.google.com/d/msg/django-developers/6j7gkJkAa9s/u5B2Ys3-DQAJ

schinckel commented:
Oh nice. I struggle to remember _all_ of the steps to get the default app config to load ;)

I can't see what the behaviour is/should be when there are multiple configs that each assert default.

Should that be an error?

aaugustin commented:
> I can't see what the behaviour is/should be when there are multiple configs that each assert default.

Currently, none of them will be used.

> Should that be an error?

Yes, it should.

I thought about that at some point and then I forgot!

GeyseR commented:
Currently, it is possible to define something like "extended" app config but not use it as a default one. With the proposed implementation, it looks like such an extended profile will be loaded silently without any warning. Shouldn't we consider this as a backward incompatibility and raise a warning?
This is just a random finding during a quick review of the code, I personally don't use such single non-default configs and don't know anyone who uses :)

schinckel commented:
I think you could implement that by detecting existing ones that don't have a `default_app_config` set, but do have exactly one config in `apps.py`

aaugustin commented:
> Currently, it is possible to define something like "extended" app config but not use it as a default one. With the proposed implementation, it looks like such an extended profile will be loaded silently without any warning. Shouldn't we consider this as a backward incompatibility and raise a warning?

Yes, I should add a deprecation warning for this. Thanks for spotting the edge case!

rsalmaso commented:
I'm the only one who want explicit config only?

Every time I propose a patch to add `AppConfig` (and relative `defautlt_app_config`) I'm asked why they should use AppConfig and/or use `default_app_config` (IE: https://github.com/django-wiki/django-nyt/pull/57).

aaugustin commented:
@rsalmaso I'd like to be sure we understand each other — the language barrier might be getting in the way.

To clarify, this PR:

- removes support for `default_app_config` (after a deprecation path). I understand that you're OK with this.
- adds "magic" detection of the app config, when there's only one reasonable choice. I understand that you disagree with this. So did I when I wrote the current implementation (which this PR changes)!

However, "magic" appears to be what most users of the framework want here. After five years, my recommendation doesn't get traction; it's time to reverse course.

auvipy commented:
what will happen to nested apps? will this change allow identifying nested apps?

aaugustin commented:
> > Currently, it is possible to define something like "extended" app config but not use it as a default one. With the proposed implementation, it looks like such an extended profile will be loaded silently without any warning. Shouldn't we consider this as a backward incompatibility and raise a warning?
> 
> Yes, I should add a deprecation warning for this. Thanks for spotting the edge case!

If I do this, taking advantage of the new feature (i.e. not declaring default_app_config and letting Django detect the default AppConfig class) will trigger the deprecation warning.

Not many apps provide an AppConfig class that isn't used by default. I'd like to document this as a backwards-incompatible change and tell the developers of such apps to add `default = False` to ensure the AppConfig isn't loaded by default.

GeyseR commented:
Hi @aaugustin 
I actually don't get an idea of these lines https://github.com/django/django/pull/12310/files#diff-239ff7e40a4b9921b462471882a575f6R124-R129
Shouldn't we just consider all of the configs as having the `default` setting equals to `True`?
This way it would be possible to have from zero to one default configs.

aaugustin commented:
>  Shouldn't we just consider all of the configs as having the `default` setting equals to `True`?

There are two reasons for the current implementation:

- An app should be able to prevent automatic detecting of `AppConfig` subclasses. This should be uncommon, but it's possible by setting `default = False` on all classes in the `apps` submodule. (Of course, you can also define `AppConfig` subclasses in another module.)
- When an app provides two or more configs, I prefer setting `default = True` on the default config than setting `default = False` on all non-default configs. This is what the code you highlighted does.

GeyseR commented:
Yea, I have read through the code one more time and finally understood it :)

> If I do this, taking advantage of the new feature (i.e. not declaring default_app_config and letting Django detect the default AppConfig class) will trigger the deprecation warning.
> Not many apps provide an AppConfig class that isn't used by default. I'd like to document this as a backwards-incompatible change and tell the developers of such apps to add default = False to ensure the AppConfig isn't loaded by default.
those are valid points (I personally don't know any cases like that) and would agree with your idea to just document the backwards-incompatible change

MarkusH commented:
I'm not sure I get why you propose to add the "magic" to auto-load an `AppConfig` when there's only one? Can't we deprecate `default_app_config` with the intention that the only valid type of string in `INSTALLED_APPS` is one that points to a subclass of `AppConfig`?

What are your reservations with requiring users to put e.g. `"my.own.library.apps.MyAppConfig"` or `"third.party.library.apps.MyAppConfig"` in `INSTALLED_APPS`? The installation instructions for other apps will only need to change in so far as that they'd need to include the path to the class, which _will_ be a burden during upgrading. But for users adding a new library to their project, it will not change anything, as they're adding the "fully quoted name" from the beginning.

jarshwah commented:
@MarkusH that's adding churn for churns sake. The entire point of this PR is that everyone is relying on apps defining `default_app_config` because users aren't adding AppConfigs in their INSTALLED_APPS.
Forcing the addition of AppConfig into `INSTALLED_APP` strings is a strictly worse user experience for all involved.

For example, my major project has about 110 installed apps. About 15 of them have explicit AppConfig labels. If a force to define the config was used, I'd need to go look up what the name of the AppConfig is for 95 projects.

jarshwah commented:
@aaugustin one pattern I'm aware of is defining the AppConfig in an apps `__init__.py` file, and directly connecting it to `default_app_config`. It's my understanding that this patch would break that pattern.

FWIW, it's not one I agree with or endorse, but major $work$ project has 16 apps defined that way.

aaugustin commented:
@MarkusH You don't need to convince _me_ that explicit references to AppConfig classes in `INSTALLED_APPS` are good (even if verbose). My intent in the app-loading refactor was to do what you're proposing, but it's been five years and we're clearly not in a position to make it happen.

Deprecating `default_app_config` would annoy almost every Django user for minimal benefits. Sometimes committers are criticized for making arbitrary changes without considering the burden for users. Deprecating `default_app_config` would clearly fall on the wrong side of this line.

The patch in this PR looks fairly maintainable — once `default_app_config` goes away and we remove about half the conditional branches — so let's make this small effort to keep things simple for users.

@jarshwah Yes, this is another common practice, but less common than `apps.py` which is demonstrated in the docs and in the app template. I'm not going to support more than one convention.

We'll tell these users to use the full path in `INSTALLED_APPS` or move the `AppConfig` to `apps.py`.

adamchainz commented:
I am in favour of this change, and don't have much else to add. Thanks for everyone's inputs to the discussion so far.

carltongibson commented on docs/ref/applications.txt:
```diff
@@ -56,25 +56,16 @@ application and have models, etc. (which would require adding it to
-module, Django checks for a ``default_app_config`` variable in that module.
+module, if Django finds exactly one :class:`~django.apps.AppConfig` subclass
+in the ``apps.py`` submodule, it uses that configuration for the application.
+Else, it uses the base :class:`~django.apps.AppConfig` class.
```
What about the `default=True`? — I think we need to be super explicit here. 

(Copying from the comment on the release notes diff):

There's a three step algorithm: 

1.  Automatic discovery of single AppConfig.
2. Finer-grained control using default.
3. Explicit class path in INSTALLED_APPS.

I think we have to spell that out here. (Then perhaps a link back from the release notes... ` See ... for more details.`)

carltongibson commented on docs/ref/applications.txt:
```diff
@@ -56,25 +56,16 @@ application and have models, etc. (which would require adding it to
-require the dotted path to the appropriate :class:`~django.apps.AppConfig`
-subclass to be configured explicitly in :setting:`INSTALLED_APPS`.
+Alternatively, :setting:`INSTALLED_APPS` may contain the dotted path to a
+configuration class.
```
Example here. (I think this can be confusing.) 

```python
INSTALLED_APPS = [
    ...
    'polls.apps.PollsAppConfig',
    ...
]
```

carltongibson commented on docs/releases/3.1.txt:
```diff
@@ -438,6 +459,12 @@ Miscellaneous
+* When an application defines an :class:`~django.apps.AppConfig` subclass in
+  an ``apps.py`` submodule, Django 3.1 uses this configuration automatically,
+  even if it isn't enabled with ``default_app_config``.  Set ``default =
+  True`` in the :class:`~django.apps.AppConfig` subclass if you need to
```
`default=False` to prevent this behavior?

carltongibson commented on docs/ref/applications.txt:
```diff
@@ -148,8 +130,12 @@ configuration::
+submodule called ``apps.py``. This is a convention, not a requirement.
+:class:`~django.apps.AppConfig` subclasses may be defined anywhere.
+
+In this situation, :setting:`INSTALLED_APPS` must contain the dotted path to
```
"In this situation" — what situation? The one in the last paragraph? (But which... the actual example, or the subjunctive `may be defined anywhere`?) 

> Where the AppConfig is defined somewhere other than in `apps.py`...

carltongibson commented on docs/releases/3.1.txt:
```diff
@@ -27,6 +27,27 @@ officially support the latest release of each series.
+pointing to this class in their ``__init__.py``.
+
+When the ``apps.py`` submodule exists and defines a single
+:class:`~django.apps.AppConfig`, Django 3.1 uses that configuration
```
> :class:`~django.apps.AppConfig` subclass, ...

carltongibson commented on docs/releases/3.1.txt:
```diff
@@ -27,6 +27,27 @@ officially support the latest release of each series.
 What's new in Django 3.1
 ========================
 
+Automatic :class:`~django.apps.AppConfig` detection
```
`s/detection/discovery/`? 

We have "discovery" elsewhere. I think one or the other (always) helps folks know we're (still) talking about the same thing.

carltongibson commented on docs/releases/3.1.txt:
```diff
@@ -27,6 +27,27 @@ officially support the latest release of each series.
+to switch the ecosystem to the latter, but the switch didn't happen.
+
+With the automatic discovery, most apps no longer need ``default_app_config``.
+As a consequence, it's deprecated.
```
It's not "most apps" right? It's all apps. 

1. Automatic discovery of single AppConfig. 
2. Finer-grained control using `default`. 
3. Explicit class path in INSTALLED_APPS.

felixxm commented on django/apps/config.py:
```diff
@@ -82,87 +87,159 @@ def _path_from_module(self, module):
+                    )
+                warnings.warn(
+                    message,
+                    RemovedInDjango40Warning,
```
`RemovedInDjango40Warning` -> `RemovedInDjango41Warning`

carltongibson replied:
Good. Eagle eye. 😀

felixxm commented on django/apps/config.py:
```diff
@@ -82,87 +87,159 @@ def _path_from_module(self, module):
-                    issubclass(candidate, AppConfig) and
-                    candidate is not AppConfig
+                message = (
+                    "%r defines default_app_config = %r. "
```
Chop trailing space.

carltongibson replied:
This is here for appending to the follow-ups later. One or other is always appended so...

felixxm replied:
True, I missed this.

felixxm commented on django/apps/config.py:
```diff
@@ -82,87 +87,159 @@ def _path_from_module(self, module):
-                        % (mod_path, cls_name, ', '.join(candidates))
+                if new_entry == app_config_name:
+                    message += (
+                        "Django now detects this configuration "
+                        "automatically. You can simply remove "
+                        "default_app_config."
```
```suggestion
                        'Django now detects this configuration automatically. '
                        'You can simply remove default_app_config.'
```

felixxm commented on docs/internals/deprecation.txt:
```diff
@@ -24,6 +24,8 @@ details on these changes.
 * The ``whitelist`` argument and ``domain_whitelist`` attribute of
   ``django.core.validators.EmailValidator`` will be removed.
 
+* The ``default_app_config`` hook will be removed.
```
I wouldn't call it `hook`, maybe `module variable` :thinking:

felixxm commented on docs/ref/applications.txt:
```diff
@@ -90,32 +94,24 @@ would provide a proper name for the admin::
+:class:`AppConfig` subclasses may be defined anywhere. The ``apps.py``
+convention merely allows Django to load them automatically when
+:setting:`INSTALLED_APPS` contains the path to an application module rather
+than the path to a configuration class.
```
`versionchanged` with a short explanation is missing.

felixxm commented on django/apps/config.py:
```diff
@@ -82,87 +87,159 @@ def _path_from_module(self, module):
-                import_module(entry)
+                else:
+                    message += (
+                        "Django detects another configuration "
+                        "automatically, %r. You should review "
+                        "your configuration and make sure Django "
+                        "can detect the default one correctly."
+                        % app_config_name
```
Wrap at 79 chars.

felixxm commented on django/apps/config.py:
```diff
@@ -82,87 +87,159 @@ def _path_from_module(self, module):
+                if new_entry == app_config_name:
+                    message += (
+                        "Django now detects this configuration "
+                        "automatically. You can simply remove "
```
I would chop `simply`.

carltongibson commented:
Thanks @felixxm — I got most of them (bar the one, but I wasn't sure of a better re-formatting...) 

Let's see if @aaugustin has time/inclination to review again, but otherwise LGTM.

aaugustin commented:
I'll take a look over the week-end. I saw your email, but didn't have time last week-end!

aaugustin commented:
Thanks for picking the flag, BTW. I didn't remember this was such a large patch. (Not huge, but not trivial either.)

carltongibson commented:
Hi @aaugustin. Thanks. There's no rush, take the time you need.

kezabelle commented on django/apps/config.py:
```diff
@@ -82,87 +87,157 @@ def _path_from_module(self, module):
-                import_module(entry)
+                else:
+                    message += (
+                        "Django detects another configuration automatically, "
```
`Django detects another configuration automatically` reads awkwardly to me. I assume it should be taken to mean Django detected another candidate to use as the default?

Or put another way - the test case (elsewhere in the PR) which shows the full output of this warning being raised doesn't tell me (as someone seeing it for the first time) what the _actual_ problem is, nor what I should to do to fix it (beyond review my configuration)

aaugustin replied:
This is a highly improbable edge case:

- app contains an `apps` submodule declaring a single `AppConfig` subclass
- app `__init__.py` contains a `default_app_config` variable pointing to another `AppConfig` subclass

I added this check and error message for the sake of completeness but I'm failing to imagine a legitimate use case for this.

Thanks for pointing out that the message isn't sufficiently helpful. I agree that we should do better. How about this?

```
"However, Django's automatic detection picked another configuration, "
"%r. You should move the default AppConfig class to the apps submodule "
" of your application and, if this module defines several AppConfig classes, "
"mark the default one with default = True."
```

kezabelle replied:
I like that it's a highly improbable edge case but you've handled it anyway :)
Yep that text (or something like it) seems clearer to me, as it's giving instruction for resolution, should it ever happen to anyone.

kezabelle commented on django/apps/config.py:
```diff
@@ -82,87 +87,157 @@ def _path_from_module(self, module):
+        # If import_module succeeds, entry points to the app module.
+        try:
+            app_module = import_module(entry)
+        except Exception:
```
Silencing far more exceptions than previously (where it was `ImportError`) ... is this intentional, and will it lead to people having a harder time debugging when their app config goes wrong? I guess it'd be somewhere higher up the chained stack traces, maybe?
Same expansion of caught exceptions (`ImportError` -> `Exception`) happens at the `import_string` so I assume the answer to this also answers that.

aaugustin replied:
All errors are ignored here (and a second time below).

Then there's logic to provide a good error message, whether the user tried the path to a module or an AppConfig class: https://github.com/django/django/pull/12310/files#diff-239ff7e40a4b9921b462471882a575f6R187-R209

Specifically, the same exception as here is re-raised there:

```python
                # Re-trigger the module import exception.
                import_module(entry)
```

aaugustin replied:
So, yes, this is intentional. It only depends on the assumption that running `import_module` twice doesn't make debugging significantly harder than running it once.

This assumption may be wrong in the presence e.g. of caching. If importing a Django application's module (not even the models module) or AppConfig class isn't idempotent, I feel like the user has been looking for trouble...

I think it's better to provide a nice error message for the common error case — typos — than to increase slightly the possibility that extremely rare and very disputable edge cases might be marginally easier to debug :-)

aaugustin commented on django/apps/config.py:
```diff
@@ -82,87 +87,157 @@ def _path_from_module(self, module):
+                if new_entry == app_config_name:
+                    message += (
+                        'Django now detects this configuration automatically. '
+                        'You can remove default_app_config.'
```
Double quotes like surrounding messages?

aaugustin commented on docs/ref/applications.txt:
```diff
@@ -56,25 +56,29 @@ application and have models, etc. (which would require adding it to
+module, by default, if Django finds exactly one :class:`AppConfig`
+subclass in the ``apps.py`` submodule, it uses that configuration for the
+application. This behavior may be controlled by setting
+:attr:`AppConfig.default` to ``False``.
```
This behavior may be **disabled**?

I find "controlled" insufficiently specific there.

aaugustin commented on docs/ref/applications.txt:
```diff
@@ -147,8 +148,13 @@ configuration::
+
+Where the AppConfig is defined somewhere other than in an ``apps.py``,
+:setting:`INSTALLED_APPS` must contain the dotted path to the configuration
+class.
```
I believe this paragraph is misleading. In this example, `anthology` is the "Django project", not a "Django app"; it isn't intended to be in `INSTALLED_APPS`. Therefore, there will be no automatic discovery from `anthology.apps`.

I would keep just the first paragraph here and remove the second one.

aaugustin replied:
Or you can revert to my previous attempt:

```
In this situation, :setting:`INSTALLED_APPS` must contain the dotted path to
the configuration class because it lives outside of an application.
```

aaugustin commented:
Hello @carltongibson — thanks for completing and improving the documentation! We're almost there.

Summarizing my comments above: there's one error message to optimize, one change to the docs to revert, and a couple minor things. To avoid more back and forth, perhaps it's best if you have the final call on the wording?

Finally, I think this should fix the failing test:

```diff
diff --git a/tests/i18n/loading_app/apps.py b/tests/i18n/loading_app/apps.py
index b4cdece232..14cfdf776f 100644
--- a/tests/i18n/loading_app/apps.py
+++ b/tests/i18n/loading_app/apps.py
@@ -2,4 +2,4 @@ from django.apps import AppConfig


 class LoadingAppConfig(AppConfig):
-    name = 'loading_app'
+    name = 'i18n.loading_app'
```

Given that `tests/i18n` wasn't on `INSTALLED_APPS`, the previous version wasn't supposed to work, and supporting it was a bug.

carltongibson commented:
Hi @aaugustin. Super. Thanks for reviewing. I will pick up those comments I the week. 

Nice work! 🏅

aaugustin commented:
To save you a bit of time I rebased and added three commits corresponding to my comments. Let's see if tests pass ;-) and feel free to adjust the wording further if needed.

felixxm commented on tests/apps/tests.py:
```diff
@@ -84,25 +90,83 @@ def test_no_such_app(self):
+        """Load an app that doesn't provide an AppConfig class."""
+        with self.settings(INSTALLED_APPS=['apps.no_config_app']):
+            config = apps.get_app_config('no_config_app')
+        self.assertEqual(type(config), AppConfig)
```
Is there a reason to not use `assertIsInstance()` (like previously) instead of `assertEqual(type(...), ...)`?
